snow/engine/snowman/bootstrap/bootstrapper.go

```diff
@@ -162,25 +176,27 @@ func (b *Bootstrapper) MultiPut(vdr ids.ShortID, requestID uint32, blks [][]byte
 			vdr, requestID)
 		return nil
 	}
-
-	wantedBlk, err := b.VM.ParseBlock(blks[0]) // the block we requested
-	if err != nil {
-		b.Ctx.Log.Debug("Failed to parse requested block %s: %s", wantedBlkID, err)
-		return b.fetch(wantedBlkID)
-	} else if actualID := wantedBlk.ID(); !actualID.Equals(wantedBlkID) {
-		b.Ctx.Log.Debug("expected the first block to be the requested block, %s, but is %s",
-			wantedBlk, actualID)
-		return b.fetch(wantedBlkID)
-	}
-
-	for _, blkBytes := range blks {
-		if _, err := b.VM.ParseBlock(blkBytes); err != nil { // persists the block
+	var blk, tail snowman.Block
+	var err error
+	for i, blkBytes := range blks {
+		if blk, err = b.VM.ParseBlock(blkBytes); err != nil { // Parse the block from bytes
 			b.Ctx.Log.Debug("Failed to parse block: %s", err)
 			b.Ctx.Log.Verbo("block: %s", formatting.DumpBytes{Bytes: blkBytes})
+			return b.fetch(wantedBlkID)
+		}
+		if i == 0 {
+			tail = blk
 		}
+		blkID := blk.ID()
+		if !blkID.Equals(wantedBlkID) {
+			b.Ctx.Log.Debug("expected the next block to be %s but is %s", wantedBlkID, blkID)
+			return b.fetch(wantedBlkID)
+		}
+		wantedBlkID = blk.Parent()
+		b.blockCache.Put(blkID, blk) // Put block in cache
```

Comment:
We should document that we are assuming the blockCache size is larger than `common.MaxContainersPerMultiPut`